lib/fined.js

```diff
@@ -0,0 +1,151 @@
+const fs = require('fs');
+const path = require('path');
+
+const isString = require('lodash.isstring');
+const isArray = require('lodash.isarray');
+const isPlainObject = require('lodash.isplainobject');
+const isEmpty = require('lodash.isempty');
+const pick = require('lodash.pick');
+
+const expandTilde = require('expand-tilde');
+const parsePath = require('parse-filepath');
+
+var Fined = function(basedir, defaultExtensions) {
+  Constructor.prototype = Fined.prototype;
+  return new Constructor(basedir, defaultExtensions);
+
+  function Constructor(basedir, exts) {
+    this.basedir = path.resolve(isString(basedir) ? basedir : '.');
+    this.defaultExtensionArray = createExtensionArray(exts);
+    this.defaultExtensionMap = createExtensionMap(exts);
+  }
+};
+
+Fined.prototype.find = function(searchPath, name) {
+  var expandedArr = expandPath.call(this, searchPath, name);
+  for (var i = 0, n = expandedArr.length; i < n; i++) {
+    var found = findWithExpandedPath(expandedArr[i]);
+    if (found) {
+      return found;
+    }
+  }
+  return null;
+};
+
+function expandPath(searchPath, name) {
+  var extArr = this.defaultExtensionArray, extMap = this.defaultExtensionMap;
+  var filePath, findUp = false;
+
+  if (isString(searchPath)) {
+    filePath = searchPath;
+
+  } else if (isPlainObject(searchPath) && isString(searchPath.path)) {
+    filePath = searchPath.path;
+    findUp = !!searchPath.findUp;
+
+    if ('extensions' in searchPath) {
+      extArr = createExtensionArray(searchPath.extensions);
+      extMap = createExtensionMap(searchPath.extensions);
+    }
+  } else {
+    return [];
+  }
+
+  filePath = expandTilde(filePath);
+
+  if (isString(name) && name.length > 0) {
+    filePath = path.join(filePath, name);
+  }
+
+  return extArr.map(createExpandedPath.bind(this));
+
+
+  function createExpandedPath(ext) {
```

Comment:
I would avoid creating functions inside of functions... just define this outside the scope of the main function body and pass in the values you need. This will make it a zillion times easier to test, as well.